aidb/vector_database/marqo_vector_database.py

```diff
@@ -0,0 +1,223 @@
+  pwd: Optional[str] = field(default=None)
+  api_key: Optional[str] = field(default=None)
+
+class MarqoVectorDatabase(VectorDatabase):
+  def __init__(self, marqo_auth: MarqoAuth, embedding_dim: int) -> None:
+    '''
+    Authentication
+    '''
+    if marqo_auth.url is None:
+      raise ValueError('Marqo requires URL to connect')
+    
+    self.marqo_client = marqo.Client(
+      url=marqo_auth.url, main_user=marqo_auth.username, main_password=marqo_auth.pwd)
+    status = self.marqo_client.get_marqo()
+    if not status:
+      raise Exception('Initial connection to Marqo failed')
+    
+    self.index_list = [i.index_name for i in self.marqo_client.get_indexes()['results']]
+    self.embedding_dim = embedding_dim
+    
+
+  @staticmethod
+  def _get_auth_secret(username: Optional[str] = None, password: Optional[str] = None, api_key: Optional[str] = None):
+    '''
+    verify the user information
+    '''
+    if api_key:
+      return api_key
+    elif username and password:
+      return (username, password)
+    else:
+      raise Exception('Please provide api or username and password to connect Marqo')
+    
+  
+  def create_index(
+    self,
+    index_name: str,
+    similarity: str = 'l2',
+    recreate_index: bool = False
+  ):
+    '''
+    Create a new index
+    :similarity: similarity function, it should be one of l1, l2, linf and cosinesiml
+    '''
+    if recreate_index:
+      if index_name in self.index_list:
+        self.delete_index(index_name)
+
+    if not index_name:
+      raise Exception('Index name must not be none')
+
+    if similarity not in ['cosinesimil', 'l1', 'l2', 'linf']:
+      raise Exception('Similarity function must be one of l1, l2, linf and cosinesiml')
+
+    if index_name in self.index_list:
+      raise Exception(f'Index {index_name} already exists, please use another name')
+
+    # Not using any model, providing own vectors
+    # https://docs.marqo.ai/1.4.0/API-Reference/Indexes/create_index/#no-model
+    self.marqo_client.create_index(
+      index_name=index_name,
+      settings_dict={
+                'index_defaults': {
+                    'model': 'no_model',
+                    'model_properties': {
+                        'dimensions': self.embedding_dim
+                    },
+                    'ann_parameters':{
+                        'space_type': similarity
+                    }
+                }
+            }
+      )
+    self.index_list.append(index_name)
+
+
+  def load_index(self):
+    '''
+    Reload index from Marqo
+    '''
+    self.index_list = [i.index_name for i in self.marqo_client.get_indexes()['results']]
+
+
+  def delete_index(self, index_name: str):
+    '''
+    delete an index
+    '''
+    if index_name in self.index_list:
+      self.marqo_client.delete_index(index_name)
+      # self.index_list.remove(index_name)
+      self.index_list = [i.index_name for i in self.marqo_client.get_indexes()['results']]
```

Comment:
Remove commented code. If you want to reload index from the vector database, you can use 'load_index'.

Reply:
Done